The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14908345553/job/41876201981) [(plain)](https://github.com/rust-lang/rust/commit/8cd43b3d5e71f93f8a47c0bfd6069231a99a682a/checks/41876201981/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/pass/float_nan.rs ... ok
tests/pass/0weak_memory_consistency_sc.rs ... ok

FAILED TEST: tests/pass/shims/x86/intrinsics-x86-sse2.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-hZNLlt" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/miri_ui/0/tests/pass/shims/x86" "tests/pass/shims/x86/intrinsics-x86-sse2.rs" "--edition" "2021"

error: test got exit status: 1, but expected 0
 = note: compilation failed, but was expected to succeed

error: no output was expected
Execute `./miri test --bless` to update `tests/pass/shims/x86/intrinsics-x86-sse2.stderr` to the actual output
+++ <stderr output>
error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:388:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`
   |
   = note: `-D unnecessary-transmutes` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(unnecessary_transmutes)]`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:397:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:406:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:415:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:424:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:433:36
   |
LL |         let e = _mm_setr_epi64x(0, transmute(2.0f64));
   |                                    ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:442:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:451:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:460:36
   |
LL |         let e = _mm_setr_epi64x(0, transmute(2.0f64));
   |                                    ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:469:36
   |
LL |         let e = _mm_setr_epi64x(0, transmute(2.0f64));
   |                                    ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:478:36
   |
LL |         let e = _mm_setr_epi64x(0, transmute(2.0f64));
   |                                    ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:487:36
   |
LL |         let e = _mm_setr_epi64x(0, transmute(2.0f64));
   |                                    ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: miri cannot be run on programs that fail compilation

error: aborting due to 13 previous errors



full stderr:
error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:388:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`
   |
   = note: `-D unnecessary-transmutes` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(unnecessary_transmutes)]`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:397:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:406:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:415:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:424:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:433:36
   |
LL |         let e = _mm_setr_epi64x(0, transmute(2.0f64));
   |                                    ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:442:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:451:37
   |
LL |         let e = _mm_setr_epi64x(!0, transmute(2.0f64));
   |                                     ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:460:36
   |
LL |         let e = _mm_setr_epi64x(0, transmute(2.0f64));
   |                                    ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:469:36
   |
LL |         let e = _mm_setr_epi64x(0, transmute(2.0f64));
   |                                    ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:478:36
   |
LL |         let e = _mm_setr_epi64x(0, transmute(2.0f64));
   |                                    ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: unnecessary transmute
##[error]  --> tests/pass/shims/x86/intrinsics-x86-sse2.rs:487:36
   |
LL |         let e = _mm_setr_epi64x(0, transmute(2.0f64));
   |                                    ^^^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0f64).cast_signed()`

error: miri cannot be run on programs that fail compilation

error: aborting due to 13 previous errors

---

Location:
   /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ui_test-0.29.2/src/lib.rs:369

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-3ef105e8c63d8dc3` (exit status: 1)
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:03:54
  local time: Thu May  8 14:33:18 UTC 2025
  network time: Thu, 08 May 2025 14:33:18 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_mir_transform/src/check_unnecessary_transmutes.rs
+++ b/compiler/rustc_mir_transform/src/check_unnecessary_transmutes.rs
@@ -55,18 +55,45 @@ impl<'a, 'tcx> UnnecessaryTransmuteChecker<'a, 'tcx> {
             },
             // char → u32
             (Char, Uint(UintTy::U32)) => err(format!("u32::from({arg})")),
+            // char (→ u32) → i32
+            (Char, Int(IntTy::I32)) => err(format!("u32::from({arg}).cast_signed()")),
             // u32 → char
             (Uint(UintTy::U32), Char) => Error {
                 sugg: format!("char::from_u32_unchecked({arg})"),
                 help: Some("consider `char::from_u32(…).unwrap()`"),
                 span,
             },
+            // i32 → char
+            (Int(IntTy::I32), Char) => Error {
+                sugg: format!("char::from_u32_unchecked(i32::cast_unsigned({arg}))"),
+                help: Some("consider `char::from_u32(i32::cast_unsigned(…)).unwrap()`"),
+                span,
+            },
             // uNN → iNN
             (Uint(ty), Int(_)) => err(format!("{}::cast_signed({arg})", ty.name_str())),
             // iNN → uNN
             (Int(ty), Uint(_)) => err(format!("{}::cast_unsigned({arg})", ty.name_str())),
+            // fNN → xsize
+            (Float(ty), Uint(UintTy::Usize)) => {
+                err(format!("{}::to_bits({arg}) as usize", ty.name_str()))
+            }
+            (Float(ty), Int(IntTy::Isize)) => {
+                err(format!("{}::to_bits({arg}) as isize", ty.name_str()))
+            }
+            // fNN (→ uNN) → iNN
+            (Float(ty), Int(..)) => err(format!("{}::to_bits({arg}).cast_signed()", ty.name_str())),
             // fNN → uNN
             (Float(ty), Uint(..)) => err(format!("{}::to_bits({arg})", ty.name_str())),
+            // xsize → fNN
+            (Uint(UintTy::Usize) | Int(IntTy::Isize), Float(ty)) => {
+                err(format!("{}::from_bits({arg} as _)", ty.name_str(),))
+            }
+            // iNN (→ uNN) → fNN
+            (Int(int_ty), Float(ty)) => err(format!(
+                "{}::from_bits({}::cast_unsigned({arg}))",
+                ty.name_str(),
+                int_ty.name_str()
+            )),
             // uNN → fNN
             (Uint(_), Float(ty)) => err(format!("{}::from_bits({arg})", ty.name_str())),
             // bool → { x8 }

--- a/src/tools/miri/tests/pass/shims/x86/intrinsics-x86-sse2.rs
+++ b/src/tools/miri/tests/pass/shims/x86/intrinsics-x86-sse2.rs
@@ -1,5 +1,6 @@
 // We're testing x86 target specific features
 //@only-target: x86_64 i686
+#![allow(unnecessary_transmutes)]
 
 #[cfg(target_arch = "x86")]
 use std::arch::x86::*;

--- a/tests/ui/transmute/unnecessary-transmutation.fixed
+++ b/tests/ui/transmute/unnecessary-transmutation.fixed
@@ -49,6 +49,10 @@ fn main() {
         //~^ ERROR
         let y: char = char::from_u32_unchecked(y);
         //~^ ERROR
+        let y: i32 = u32::from('🐱').cast_signed();
+        //~^ ERROR
+        let y: char = char::from_u32_unchecked(i32::cast_unsigned(y));
+        //~^ ERROR
 
         let x: u16 = i16::cast_unsigned(8i16);
         //~^ ERROR
@@ -70,6 +74,11 @@ fn main() {
         let y: f64 = f64::from_bits(3u64);
         //~^ ERROR
         let y: u64 = f64::to_bits(2.0);
+        //~^ ERROR
+
+        let y: f64 = f64::from_bits(i64::cast_unsigned(1i64));
+        //~^ ERROR
+        let y: i64 = f64::to_bits(1f64).cast_signed();
         //~^ ERROR
 
         let z: bool = (1u8 == 1);

--- a/tests/ui/transmute/unnecessary-transmutation.rs
+++ b/tests/ui/transmute/unnecessary-transmutation.rs
@@ -49,6 +49,10 @@ fn main() {
         //~^ ERROR
         let y: char = transmute(y);
         //~^ ERROR
+        let y: i32 = transmute('🐱');
+        //~^ ERROR
+        let y: char = transmute(y);
+        //~^ ERROR
 
         let x: u16 = transmute(8i16);
         //~^ ERROR
@@ -70,6 +74,11 @@ fn main() {
         let y: f64 = transmute(3u64);
         //~^ ERROR
         let y: u64 = transmute(2.0);
+        //~^ ERROR
+
+        let y: f64 = transmute(1i64);
+        //~^ ERROR
+        let y: i64 = transmute(1f64);
         //~^ ERROR
 
         let z: bool = transmute(1u8);

--- a/tests/ui/transmute/unnecessary-transmutation.stderr
+++ b/tests/ui/transmute/unnecessary-transmutation.stderr
@@ -154,82 +154,108 @@ LL |         let y: char = transmute(y);
    = help: consider `char::from_u32(…).unwrap()`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:53:22
+  --> $DIR/unnecessary-transmutation.rs:52:22
+   |
+LL |         let y: i32 = transmute('🐱');
+   |                      ^^^^^^^^^^^^^^^ help: replace this with: `u32::from('🐱').cast_signed()`
+
+error: unnecessary transmute
+  --> $DIR/unnecessary-transmutation.rs:54:23
+   |
+LL |         let y: char = transmute(y);
+   |                       ^^^^^^^^^^^^ help: replace this with: `char::from_u32_unchecked(i32::cast_unsigned(y))`
+   |
+   = help: consider `char::from_u32(i32::cast_unsigned(…)).unwrap()`
+
+error: unnecessary transmute
+  --> $DIR/unnecessary-transmutation.rs:57:22
    |
 LL |         let x: u16 = transmute(8i16);
    |                      ^^^^^^^^^^^^^^^ help: replace this with: `i16::cast_unsigned(8i16)`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:55:22
+  --> $DIR/unnecessary-transmutation.rs:59:22
    |
 LL |         let x: i16 = transmute(x);
    |                      ^^^^^^^^^^^^ help: replace this with: `u16::cast_signed(x)`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:57:22
+  --> $DIR/unnecessary-transmutation.rs:61:22
    |
 LL |         let x: u32 = transmute(4i32);
    |                      ^^^^^^^^^^^^^^^ help: replace this with: `i32::cast_unsigned(4i32)`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:59:22
+  --> $DIR/unnecessary-transmutation.rs:63:22
    |
 LL |         let x: i32 = transmute(x);
    |                      ^^^^^^^^^^^^ help: replace this with: `u32::cast_signed(x)`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:61:22
+  --> $DIR/unnecessary-transmutation.rs:65:22
    |
 LL |         let x: u64 = transmute(7i64);
    |                      ^^^^^^^^^^^^^^^ help: replace this with: `i64::cast_unsigned(7i64)`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:63:22
+  --> $DIR/unnecessary-transmutation.rs:67:22
    |
 LL |         let x: i64 = transmute(x);
    |                      ^^^^^^^^^^^^ help: replace this with: `u64::cast_signed(x)`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:66:22
+  --> $DIR/unnecessary-transmutation.rs:70:22
    |
 LL |         let y: f32 = transmute(1u32);
    |                      ^^^^^^^^^^^^^^^ help: replace this with: `f32::from_bits(1u32)`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:68:22
+  --> $DIR/unnecessary-transmutation.rs:72:22
    |
 LL |         let y: u32 = transmute(y);
    |                      ^^^^^^^^^^^^ help: replace this with: `f32::to_bits(y)`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:70:22
+  --> $DIR/unnecessary-transmutation.rs:74:22
    |
 LL |         let y: f64 = transmute(3u64);
    |                      ^^^^^^^^^^^^^^^ help: replace this with: `f64::from_bits(3u64)`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:72:22
+  --> $DIR/unnecessary-transmutation.rs:76:22
    |
 LL |         let y: u64 = transmute(2.0);
    |                      ^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(2.0)`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:75:23
+  --> $DIR/unnecessary-transmutation.rs:79:22
+   |
+LL |         let y: f64 = transmute(1i64);
+   |                      ^^^^^^^^^^^^^^^ help: replace this with: `f64::from_bits(i64::cast_unsigned(1i64))`
+
+error: unnecessary transmute
+  --> $DIR/unnecessary-transmutation.rs:81:22
+   |
+LL |         let y: i64 = transmute(1f64);
+   |                      ^^^^^^^^^^^^^^^ help: replace this with: `f64::to_bits(1f64).cast_signed()`
+
+error: unnecessary transmute
+  --> $DIR/unnecessary-transmutation.rs:84:23
    |
 LL |         let z: bool = transmute(1u8);
    |                       ^^^^^^^^^^^^^^ help: replace this with: `(1u8 == 1)`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:77:21
+  --> $DIR/unnecessary-transmutation.rs:86:21
    |
 LL |         let z: u8 = transmute(z);
    |                     ^^^^^^^^^^^^ help: replace this with: `(z) as u8`
 
 error: unnecessary transmute
-  --> $DIR/unnecessary-transmutation.rs:82:21
+  --> $DIR/unnecessary-transmutation.rs:91:21
    |
 LL |         let z: i8 = transmute(z);
    |                     ^^^^^^^^^^^^ help: replace this with: `(z) as i8`
 
-error: aborting due to 32 previous errors
+error: aborting due to 36 previous errors
 

